tests/test_restriction.cpp

```diff
@@ -0,0 +1,288 @@
+                adjacent[1] = 0.0;
+                adjacent[2] = 0.0;
+            }
+            if (j == cr_int) { //exterior-most circle- nodes to the right disappear
+                adjacent[5] = 0.0;
+                adjacent[6] = 0.0;
+                adjacent[7] = 0.0;
+            }
+
+            int k = 0;
+            std::vector<std::tuple<double, double>> h_p(8, {-1, -1}); // (h_p, h_{p-1}) relative grid sizes
+            std::vector<std::tuple<double, double>> k_q(8, {-1, -1}); // (k_q, k_{q-1})
+            // z and y represent relative positions to the coarse node. e.g. if (z,y)=(-1,1) then we consider the fine node to the top-left
+            for (int z = -1; z < 2; z++) {
+                for (int y = -1; y < 2; y++) {
+                    if (z != 0 || y != 0) {
+                        if (adjacent[k] != 0.0) {
+
+                            adjacent[k] =
+                                u_test[(2 * j + z) * p_level.ntheta_int +
+                                       ((i != 0 || y > -1)
+                                            ? (2 * i + y)
+                                            : p_level.ntheta_int -
+                                                  1)]; //adjacent value. consider periodic boundary in theta direction
+
+                            h_p[k] = {p_level.r[2 * j + z + 1] - p_level.r[2 * j + z],
+                                      p_level.r[2 * j + z] -
+                                          p_level.r[2 * j + z - 1]}; //distance in r coordinate for the adjacent node
+
+                            //to calculate k_q,k_{q-1} we consider the special case i=0 (2*i+y=y) separately.
+                            if (i > 0) {
+                                double indx =
+                                    (2 * i + y < p_level.ntheta_int - 1) ? p_level.theta[2 * i + y + 1] : 2 * PI;
+
+                                k_q[k] = {indx - p_level.theta[2 * i + y],
+                                          p_level.theta[2 * i + y] - p_level.theta[2 * i + y - 1]};
+                            }
+                            else {
+                                switch (
+                                    y) { //based on the value of y and periodic boundary conditions we get different values for {k_q,k_{q-1}}
+                                case 1:
+                                    k_q[k] = {p_level.theta[2] - p_level.theta[1], p_level.theta[1] - p_level.theta[0]};
+                                case 0:
+                                    k_q[k] = {p_level.theta[1] - p_level.theta[0],
+                                              2 * PI - p_level.theta[p_level.ntheta_int - 1]};
+                                default:
+                                    k_q[k] = {2 * PI - p_level.theta[p_level.ntheta_int - 1],
+                                              p_level.theta[p_level.ntheta_int - 1] -
+                                                  p_level.theta[p_level.ntheta_int - 2]};
```

Comment:
I dont like switch but if you want to do so, you should be consistent and I think you should not mix with if/else. Can you replace by if/else?

Reply:
done! there was also a break missing after each switch...